### PR TITLE
[0.20] [Arch] connections between stair flight and landing

### PR DESCRIPTION
**Pull request by three-ddd**
Added some features to allow the user to change the types of connections between the flight of stairs and landing.
This issue is discussed in the forum thread: https://forum.freecadweb.org/viewtopic.php?f=22&t=53018

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [ ] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Review by hyarion — CHANGES_REQUESTED** [review]
Due to the syntax error, the tests fails.
After fixing the issue, please run `./bin/FreeCAD --run-test 0` and verify that you get no errors.

**Comment by three-ddd** [comment]
Thanks for your support, hyarion.
Done.

File: src/Mod/Arch/ArchStairs.py [hunk 1]
```diff
@@ -114,7 +116,7 @@ def setProperty(obj,length,width,height,steps,name):
             elif steps:
                 stepsI = steps
             else:
-                stepsI = 16
+                stepsI = 20
```
Comment (hyarion) [inline src/Mod/Arch/ArchStairs.py:119]: May I ask why this change has been made?
Why is 20 a better number than 16?

**Comment by hyarion** [comment]
It now runs and there are no unit test errors ✅ 
My arch knowledge is close to non-existent so someone else needs to take it from here.

**Comment by paullee0** [comment]
@three-ddd I have a glance the overall changes without looking at details, looks great !

Note that your changes add the connection details to the Flight of Stairs and remove the 'projection' on the Landing.

Another idea is to add another parameter to the Landing whether to have that triangular connection details on the Landing.  So user can have even more control on each 'component.  Maybe add an issue to the bugtracker ?

Thanks !

**Comment by paullee0** [comment]
Tested a simple stairs with 3 flights, 2 landings and works great - [RC Stairs Macro](https://forum.freecadweb.org/viewtopic.php?f=22&t=53018&p=470151#p470151)

**Comment by three-ddd** [comment]
Hi, paullee thank you for taking the time to check out my work, I'm grateful to you.
As you already know, I  attempted my first PR....but something went wrong!
As you can see GitHub reports that the test is failed.....and I still don't understand why.
Can someone please help me solve this problem? Many thanks.

Note: I developed code directly on GitHub account, so I have not worked locally on my computer.

My simple workflow on Github has been:
- I created a branch called "archstairs_connections_flight_landing", in my personal repo.
- In this new branch, I modified the ArchStairs.py file
- Last, I submitted the pull request

**Comment by paullee0** [comment]
> As you already know, I attempted my first PR....but something went wrong!
> As you can see GitHub reports that the test is failed.....and I still don't understand why.

I am also python and Github begginer :D 

It seems you should have solved by following @hyarion's finding and all checks have passed :D

Maybe wait @yorikvanhavre or other core developer to further review and merge.

**Comment by yorikvanhavre** [comment]
Okay everything seems fine now, and code looks good, I can merge. Do you mind just replying @hyarion 's question, why 20 is better than 15, @three-ddd ?

**Comment by three-ddd** [comment]
Yorik, thanks for the positive feedback and I really thank you for your awesome work on FC.

@hyarion sorry for the late reply.......I chose 20 as initial value because:

    3000 (preset value of "StairsHeight") / 20 = 150 mm < 180 mm (maximum riser height)

instead:

    3000 / 16 = 187.5 mm > 180 mm (maximum riser height)


range_riser_height_stairs_colored resize

**Comment by hyarion** [comment]
> @hyarion sorry for the late reply.......

No worries :)

> I chose 20 as initial value because:
> ```
> 3000 (preset value of "StairsHeight") / 20 = 150 mm < 180 mm (maximum riser height)
> ```
> instead:
> ```
> 3000 / 16 = 187.5 mm > 180 mm (maximum riser height)
> ```

Well explained @three-ddd, thanks! 👍

**Comment by paullee0** [comment]
@three-ddd Indeed 20 steps better for default 3000m high staircase.

**Comment by three-ddd** [comment]
Hi @yorikvanhavre, 
some bugs were found, about:

- "Riser Thickness", when its value  > 0 the connection's result is not geometrically correct.

- when you set property "Structure" to "One Stringer" or "Two Stringers" options, the stair geometry fails.

Reference discussion [here](https://forum.freecadweb.org/viewtopic.php?f=22&t=53018&start=20)


> @three-ddd Indeed 20 steps better for default 3000m high staircase.

Thanks @paullee0!

**Comment by three-ddd** [comment]
> Well explained @three-ddd, thanks!

you are welcome @hyarion!

**Comment by yorikvanhavre** [comment]
@three-ddd I think this is good for merge now.. Since this modifies functionality and we are in feature-freeze mode, I'll mark it to postpone to after the 0.19 release. Could you remove the "draft" status when you feel ready?
About the 2 bugs you found, I guess there are many more... The stair tool needs a lot of work still :)

**Comment by three-ddd** [comment]
Thank Yorik.....you are right the stairs tool needs a lot of work yet ;)

**Comment by paullee0** [comment]
Great!  Look forward to more contribution from @three-ddd :D

**Comment by paullee0** [comment]
@three-ddd Could you mark your commit is merged ?  :D

**Comment by three-ddd** [comment]
> Great! Look forward to more contribution by @three-ddd :D

Thanks paullee, I hope I shall not disappoint your expectations! :D


> @three-ddd Could you mark your commit is merged ?  :D

How can I do that?

**Comment by paullee0** [comment]
Just edit your 1st post in FreeCAD forum and add e.g. [Merge] in the begining of the topic :)

**Comment by three-ddd** [comment]
> Just edit your 1st post in FreeCAD forum and add e.g. [Merge] in the begining of the topic :)

Thanks, @paullee0, for your speedy response!

**Comment by three-ddd** [comment]
> Just edit your 1st post in FreeCAD forum and add e.g. [Merge] in the begining of the topic :)

Hi @paullee0, I tried that, but it did change only title of my first post :(

first_post

**Pull request merged**